### PR TITLE
Add support for  input methods as Chinese Simplified Microsoft Pinyin and Microsoft Wubi

### LINKED ISSUE #4011: The input method candidate list can not be announce in windows 8.1
**Reported by vgjh2005 on 2014-03-22 03:50**
Hi:
My input methhd is microsoft pinyin. When I make some Inputing, NVDA cannot announce the candidate list. It can let me know which character I have typed, but I don't know which candidate is focused.
Thanks!

### LINKED ISSUE #8343: Cannot read Chinese IME candidates since Win10 updated to 1709
### Steps to reproduce:
1. Make sure Windows 10 is at 1709 edition or newer
2. Open "Language Preferences"
3. Add a Chinese language, eg: Chinese (Traditional, HKSAR)
4. Open "Options" for the newly added language
5. Add a keyboard (Chinese IME), eg: Microsoft Quick
6. Open "Notepad" application
7. Use the Chinese IME to compose text, eg: Enter "eb" then press <Space>
8. A list of candidates will be generated by the IME

### Expected behavior:
NVDA should read back the list of candidates

### Actual behavior:
NVDA is unable to read the list of candidates

### System configuration:
NVDA version: 2018.1 portable
Windows 10 is running in a VM

Windows version:
Windows 10 Home 1803 17134.48

### Other questions:
This issue has been happening on all Windows 10 PC since upgrading to 1709 edition.  The problem cannot be fixed by rebooting the PC.

NVDA 2017.4 also exhibited the same problem.

### LINKED ISSUE #8562: NVDA cannot announce candidate item in default Chinese Input method Microsoft pinyin
### Steps to reproduce:
Install  Chinese Language pack
Reboot to load language pack 
Open Notepad
Press Windows + Space  switch Input Method to Chinese Simplified China Microsoft pinyin
press h and e
then press down arrow several times
### Actual behavior:
NVDA keeps speaking he 
### Expectedhavior: 
Candidate item is announced 
If you use english espeak it would be "Chinese letter"
#### NVDA Installed/portable/running from source:
installed
#### NVDA version:
2018.2.1
#### Windows version:
Windows 8.1
#### Name and version of other software in use when reproducing the issue:
Windows 10 and 7 also has this issue 
### Other questions:
Narrator can read this.
NVDA uses ITfUIElementSink to get candidate item
 Is it possible to get candidate item from UIAutomation for Microsoft pinyin?
#### Does the issue still occur after restarting your PC?
Yes
#### Have you tried any other versions of NVDA?
alpha-15682,48201d01
This version also has this issue

### PR DESCRIPTION
**Pull request by larry801**
### Link to issue number:
Fixes #8343
Fixes #4011
Fixes #8562 
### Summary of the issue:
Since Windows 8 Microsoft changed the UI of its default Chinese input method.
This fix proposed in 2015 was not functioning properly due to UIA events issue.
Now these UIA issues have gone.
### Description of how this pull request fixes the issue:
Add support for  input methods as Chinese Simplified Microsoft Pinyin and Microsoft Wubi
through the Modern IME candidate list UI.
### Testing performed:
I built a try build then tested on several systems.
All worked perfectly
These systems are:
Windows 8.1 Enterprise
WIndows 10 Pro 1803 17735.1000 (rs5 fast ring)
Windows 10 Pro 1803 17134.191
### Known issues with pull request:
None
### Change log entry:
None

### CODE REVIEW

**Comment by josephsl** [comment]
CC @MichaelDCurran

File: source/IAccessibleHandler.py [hunk 1]
```diff
@@ -781,7 +784,8 @@ def processDestroyWinEvent(window,objectID,childID):
 	# so can't use generic focus correction. (#2695)
 	focus=api.getFocusObject()
 	from NVDAObjects.IAccessible.mscandui import BaseCandidateItem
-	if objectID==0 and childID==0 and isinstance(focus,BaseCandidateItem) and window==focus.windowHandle and not eventHandler.isPendingEvents("gainFocus"):
+	windowClassName=winUser.getClassName(window)
```
Comment (michaelDCurran) [inline source/IAccessibleHandler.py]: I don't think this gets used?

File: source/IAccessibleHandler.py [hunk 2]
```diff
@@ -781,7 +784,8 @@ def processDestroyWinEvent(window,objectID,childID):
 	# so can't use generic focus correction. (#2695)
 	focus=api.getFocusObject()
 	from NVDAObjects.IAccessible.mscandui import BaseCandidateItem
-	if objectID==0 and childID==0 and isinstance(focus,BaseCandidateItem) and window==focus.windowHandle and not eventHandler.isPendingEvents("gainFocus"):
+	windowClassName=winUser.getClassName(window)
+	if childID==0 and isinstance(focus,BaseCandidateItem) and window==focus.windowHandle and not eventHandler.isPendingEvents("gainFocus"):
```
Comment (michaelDCurran) [inline source/IAccessibleHandler.py]: I'm assuming there are objectIDs other than 0 you are seeing? can you mention what they are? Certainly negative objectIDs should never be supported here as they would be for non-client parts of the window such as the system menu or scroll bars. Were you seeing positive objectIDs?

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
On the whole this looks good. Has this been tested with languages such as Japanese or Corean? does this break any other languages or cause double speaking?

**Comment by michaelDCurran** [comment]
It may be true that some of those languages don't currently use that 
candidate UI in NvDA's code. But it is possible that they have  
previously exposed support through this new UI, but we never used it 
because it used to be broken.

Thus when we enable this new code, we could get double speaking due to 
two lots of code running.

It is highly unlikely, but I'd just like confirmation from some other 
developers who use Chinese/Japanese/Corean.

**Comment by josephsl** [comment]
Hi, I’ll report back once I locate this PR build and test it with Korean (Korean is mine). Thanks.

**Comment by josephsl** [comment]
Hi,

From what I can tell, no difference when I type using Korean in Windows 10/Notepad. Thanks.

**Comment by nishimotz** [comment]
Should I test pr8621-15831 build regarding Japanese?

**Comment by nishimotz** [comment]
I have tested as follows

* nvda_snapshot_pr8621-15831,0342aefe.exe
* Windows 10 (1803) + Microsoft IME Japanese + Microsoft Pinyin (Chinese China mainland)
* notepad

Regarding Japanese IME, no difference observed.
I think is is because Japanese IME does not have Modern UI so far.

Regarding Chinese Pinyin, I cannot read or write the language, however

* Once a Latin character is typed, candidate window pops up, which contains several candidate items.
* When NVDA is not running, I should type down arrow key to move the focus to the first candidate item. After that, down arrow key can be used to choose items.
* When this build of NVDA is running, during the type of first or second Latin character, first candidate item is focused and announced as the character descriptions.
* However, with this build, system focus is not returned to the edit control correctly, after the selection of an item, or the cancel the composition by pressing the Escape key. For example, after the cancelling of composition by Escape, pressing number characters such as 1 2 3 is not announced while they are actually typed within the notepad.

NVDAJP has worked around this issue with another approach many years ago, which should be reworked in the future.
It is nice to have this automatic focus move regarding Japanese IME as well.
However, the current pull request is not enough, I think.

**Comment by nishimotz** [comment]
Sorry for the off topic comment.
Regarding the first candidate of Japanese IME, it is not actually the same.
Issue of Japanese IME is #2730

**Comment by nishimotz** [comment]
@larry801 No difference for Japanese IME.
We should use another issue such as #2730 for details.

File: source/IAccessibleHandler.py (hunk 1 above)
Comment (larry801) [reply source/IAccessibleHandler.py]: I removed this line.

File: source/IAccessibleHandler.py (hunk 2 above)
Comment (larry801) [reply source/IAccessibleHandler.py]: Object ID of candidate item is positive. Their window handle is also different from host application.

File: source/IAccessibleHandler.py [hunk 3]
```diff
@@ -781,7 +784,7 @@ def processDestroyWinEvent(window,objectID,childID):
 	# so can't use generic focus correction. (#2695)
 	focus=api.getFocusObject()
 	from NVDAObjects.IAccessible.mscandui import BaseCandidateItem
-	if objectID==0 and childID==0 and isinstance(focus,BaseCandidateItem) and window==focus.windowHandle and not eventHandler.isPendingEvents("gainFocus"):
+	if objectID == 0 and childID==0 and isinstance(focus,BaseCandidateItem) and not eventHandler.isPendingEvents("gainFocus"):
```
Comment (michaelDCurran) [inline source/IAccessibleHandler.py]: You are now matching any destroy event with an objectID of 0 and childID of 0. This is not good.  It needs to be somehow limited to the candidate window.
Previously  it checked if it was the same window. but now you are saying it is different? In that case, what window is it exactly? I.e. is it a descendant of the focus window, or perhaps it has a specific window class?
I think it is more likely that the modern candidate UI is simply not firing a destroy event at all, and perhaps your code is matching on some other random one?

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

File: source/IAccessibleHandler.py (hunk 3 above)
Comment (larry801) [reply source/IAccessibleHandler.py]: I found a specific window class  CiceroUIWndFrame

**Comment by larry801** [comment]
I added a line to get information about events triggered after input compositon ends.
```
log.info(u"Class Name:{0}\nHWND:{1}\nFocus:{2}\nFocusHWND{3}\n{4}\n{5}".format(windowClassName,window,focus,focus.windowHandle,focus.container.windowHandle,focus.parent.windowHandle))
```
Then I got these logs.
```
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.532):
Class Name:CiceroUIWndFrame
HWND:526052
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.542):
Class Name:CiceroUIWndFrame
HWND:395012
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.552):
Class Name:MSCTFIME Composition
HWND:1247086
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.562):
Class Name:MSCTFIME Composition
HWND:1836900
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.572):
Class Name:MSCTFIME Composition
HWND:1967706
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
INFO - IAccessibleHandler.processDestroyWinEvent (15:10:33.582):
Class Name:MSCTFIME Composition
HWND:1377950
Focus:<baseObject.Dynamic_ButtonModernCandidateUICandidateItemIAccessible object at 0x065AA770>
FocusHWND1181528
526362
1181528
```
Cicero is internal name of TSF. It should be the right window class name.
I tested on 
Windows 8.1 Enterprise
Windows 10 Pro 1803 17134.191
In Microsoft Pinyin or Wubi, system focus returns correctly now.
@nishimotz  @josephsl 
Would you please test the new build against Korean and Japanese IME?

**Comment by timothytylee** [comment]
This patch does not appear to fix bug #8343.

Just for reference, MSDN has a page on TSF (https://docs.microsoft.com/en-us/previous-versions/windows/apps/hh967425(v=win.10)#accessibility), and in the section regarding on IME accessibility, it indicates that candidate windows should have a UIA automation ID of "IME_Candidate_Window" or "IME_Predication_Window".

Perhaps those automation ID can be used to identify candidate list?

**Comment by larry801** [comment]
Current Implementation uses IAccessible.

**Comment by larry801** [comment]
The candidate list of Microsoft quick has a different structure. That is why this patch will not work.
cc @timothytylee

**Comment by larry801** [comment]
In nvda find dialog under browse mode,focus correction does not work.
I recorded events with accevent.exe and found that
Window destroy events generated in nvda.exe only can be caught in "out of context" mode.
If record events without screen reader  all of these events are "out of context".
I wonder how NVDA proxies them as "in context".

**Comment by dingpengyu** [comment]
I hope this pr can be tested in the alpha version early.

**Comment by larry801** [comment]
ChangeLog Entry
Added support for Chinese input methods (Microsoft Pinyin ,Microsoft Wubi Microsoft Quick and Microsoft ChangJie) in Windows 8.1 and Windows 10

File: source/IAccessibleHandler.py [hunk 4]
```diff
@@ -569,6 +569,9 @@ def winEventCallback(handle,eventID,window,objectID,childID,threadID,timestamp):
 				window=tempWindow
 
 		windowClassName=winUser.getClassName(window)
+		# Modern IME candidate list windows fire menu events which confuse us and can't be used properly inconjunction with input composition support
```
Comment (leonardder) [inline source/IAccessibleHandler.py]: This line is pretty long. Could you please split it?

```suggestion
		# Modern IME candidate list windows fire menu events which confuse us
		# and can't be used properly in conjunction with input composition support.
```

File: source/NVDAHelper.py [hunk 5]
```diff
@@ -159,9 +159,22 @@ def handleInputCompositionEnd(result):
...
-		#Candidate list is still up
-		curInputComposition=focus.parent
-		focus.parent=focus.parent.parent
+		# Correct focus for ModernCandidateUICandidateItem
+		if isinstance(focus ,ModernCandidateUICandidateItem):
+			obj=focus.container
+			curInputComposition = obj
+			if obj:
+				if isinstance(obj,InputComposition):
+					obj=obj.parent
```
Comment (leonardder) [inline source/NVDAHelper.py]: Could you add a comment that explains why you're traversing to the parent here?

File: source/NVDAObjects/IAccessible/__init__.py [hunk 6]
```diff
@@ -457,7 +457,7 @@ def findOverlayClasses(self,clsList):
 			parentWindow=winUser.getAncestor(self.windowHandle,winUser.GA_PARENT)
 			if parentWindow and winUser.getClassName(parentWindow)=="Frame Notification Bar":
 				clsList.append(IENotificationBar)
-		if windowClassName.lower().startswith('mscandui'):
+		if windowClassName.lower().startswith('mscandui') or windowClassName in ("Microsoft.IME.CandidateWindow.View","Microsoft.IME.UIManager.CandidateWindow.Host"):
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/__init__.py]: Please split this as well

```suggestion
		if (
			windowClassName.lower().startswith('mscandui')
			or windowClassName in (
				"Microsoft.IME.CandidateWindow.View",
				"Microsoft.IME.UIManager.CandidateWindow.Host"
		)):
```

File: source/NVDAObjects/IAccessible/mscandui.py [hunk 7]
```diff
@@ -213,10 +213,48 @@ def event_show(self):
 		item=MSCandUIWindow_candidateListItem(IAccessibleObject=self.IAccessibleObject,IAccessibleChildID=3)
 		reportSelectedCandidate(item)
 
+class ModernCandidateUICandidateItem(BaseCandidateItem):
+
+	def _get_parent(self):
+		# Candidate list in Microsoft Quick cannot be obtained in IAccessible _get_parent. Use _get_parent in NVDAObject.window
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/mscandui.py]: Please split this

```suggestion
		# Candidate list in Microsoft Quick cannot be obtained in IAccessible _get_parent.
		# Use _get_parent in NVDAObject.window.
```

File: source/NVDAObjects/IAccessible/mscandui.py [hunk 8]
```diff
@@ -213,10 +213,48 @@ def event_show(self):
 		item=MSCandUIWindow_candidateListItem(IAccessibleObject=self.IAccessibleObject,IAccessibleChildID=3)
 		reportSelectedCandidate(item)
 
+class ModernCandidateUICandidateItem(BaseCandidateItem):
+
+	def _get_parent(self):
+		# Candidate list in Microsoft Quick cannot be obtained in IAccessible _get_parent. Use _get_parent in NVDAObject.window
+		parent=super(IAccessible,self).parent
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/mscandui.py]: Mmm. You're saying that you want to call _get_parent on Window, but you are calling super. I know that it is effectively the same, but I'd rather have you call Window instead if you want to make sure that this is not ambiguous. Note that you have to add an import.

```suggestion
		parent=NVDAObjects.window.Window._get_parent(self)
```

File: source/NVDAObjects/IAccessible/mscandui.py [hunk 9]
```diff
@@ -213,10 +213,48 @@ def event_show(self):
...
+		return parent
+
+	def _get_candidateCharacters(self):
+		return super(BaseCandidateItem,self).name
+
+	_candidateNumber=""
+
+	def _get_candidateNumber(self):
+		# The following property call sets candidateNumber
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/mscandui.py]: In the current code, this is only the case the first time you're calling this. Is that expected and correct?

File: source/NVDAObjects/IAccessible/mscandui.py [hunk 10]
```diff
@@ -213,10 +213,48 @@ def event_show(self):
...
+		self.visibleCandidateItemsText
+		return self._candidateNumber
+
+	def _get_visibleCandidateItemsText(self):
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/mscandui.py]: I'm not very happy with the approach where getting a property sets properties on children. This makes the code a bit vague. In my opinion, neither a getter nor a setter should touch the value of other attributes. Would you be able to come up with something that works around this?

File: source/NVDAObjects/IAccessible/mscandui.py [hunk 11]
```diff
@@ -213,10 +213,48 @@ def event_show(self):
...
+
+	def event_stateChange(self):
+		if controlTypes.STATE_SELECTED in self.states:
+			reportSelectedCandidate(self)
+
 def findExtraOverlayClasses(obj,clsList):
 	windowClassName=obj.windowClassName
 	role=obj.IAccessibleRole
-	if windowClassName=="MSCandUIWindow_Candidate":
+	if windowClassName=="Microsoft.IME.CandidateWindow.View":
```
Comment (leonardder) [inline source/NVDAObjects/IAccessible/mscandui.py]: Could you join this and the if statement beneath it in just one statement, thereby taking line length into account?

File: source/_UIAHandler.py [hunk 12]
```diff
@@ -45,6 +45,7 @@
 ]
 
 badUIAWindowClassNames=[
+	"Microsoft.IME.CandidateWindow.View",
```
Comment (leonardder) [inline source/_UIAHandler.py]: Please add a comment here that explains why this UIA implementation is considered bad.

**Review by LeonarddeR — CHANGES_REQUESTED** [review]
I'm not yet known to this area of the code, so my comments are only aiming at improving the overall code style and consistency.

File: source/_UIAHandler.py (hunk 12 above)
Comment (larry801) [reply source/_UIAHandler.py]: The code here were written by 
 @michaelDCurran 
 at https://github.com/nvaccess/nvda/commit/dbbaacea10ac1ec8c07dbe943535d86284546861
I do not know much about its meaning. 
But removing this line then candidate report will not work.
May be there are some confusing UIA events should be suppressed?

**Comment by larry801** [comment]
@leonardder 
I finally come up with a way to get candidate number without touching other objects.But now every candidate object has to traversal the whole list to get its number. After comparing timeSinceLastInput there is no noticeable difference. Now all review comments are addressed.

**Comment by dingpengyu** [comment]
Can this pr currently enter NVDA2019.3?
cc
@michaelDCurran 
and
@josephsl

**Review by michaelDCurran — COMMENTED** [review]
I have merged master and fixed up line endings of various files.
The diff sshown for this pr on Github now looks much better in that it is no longer showing several entire files changed.
Were you using git on a unix system?

**Review by michaelDCurran — APPROVED** [review]

**Comment by michaelDCurran** [comment]
@leonardder were all your comments addressed? or are you still waiting on changes.

File: source/_UIAHandler.py [hunk 13]
```diff
@@ -45,6 +45,8 @@
 ]
 
 badUIAWindowClassNames=[
+	# UIA event of candidate window interfere with MSAA events.
```
Comment (leonardder) [inline source/_UIAHandler.py]:
```suggestion
	# UIA events of candidate window interfere with MSAA events.
```

**Review by LeonarddeR — APPROVED** [review]
Just one cosmetic thing in a comment

**Comment by larry801** [comment]
@michaelDCurran 
Line ending problem may come from  my switching between source tree  and git shell.
Thanks for your help. 
@leonardder 
Now all comments have been addressed. Is this ready for merging?

**Pull request merged**